File: rust-worker/src/db/progress.rs
use serde::{Deserialize, Serialize};

/// Job stage enum
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum JobStage {
    Queue,
    Downloading,
    Whitelist,
    Generation,
    Completed,
}

impl Default for JobStage {
    fn default() -> Self {
        Self::Queue
    }
}

/// Source download status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum SourceStatus {
    Pending,
    Downloading,
    Processing,
    Completed,
    Failed,
}

impl Default for SourceStatus {
    fn default() -> Self {
        Self::Pending
    }
}

/// Format generation status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum FormatStatus {
    Pending,
    Generating,
    Compressing,
    Completed,
}

impl Default for FormatStatus {
    fn default() -> Self {
        Self::Pending
    }
}

/// Progress for a single source
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct SourceProgress {
    pub id: String,
    pub name: String,
    pub url: String,
    pub status: SourceStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_hit: Option<bool>,
    pub bytes_downloaded: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bytes_total: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub download_percent: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub download_time_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain_count: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain_change: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub warnings: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub started_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,
}

/// Whitelist pattern match info
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WhitelistPatternMatch {
    pub pattern: String,
    pub pattern_type: String,
    pub match_count: u64,
    #[serde(default)]
    pub samples: Vec<String>,
}

/// Whitelist processing progress
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct WhitelistProgress {
    pub domains_before: u64,
    pub domains_after: u64,
    pub total_removed: u64,
    pub processing: bool,
    #[serde(default)]
    pub patterns: Vec<WhitelistPatternMatch>,
}

/// Single format generation progress
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FormatProgress {
    pub format: String,
    pub status: FormatStatus,
    pub domains_written: u64,
    pub total_domains: u64,
    pub percent: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_size: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gz_size: Option<u64>,
}

/// Generation stage progress
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct GenerationProgress {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_format: Option<String>,
    #[serde(default)]
    pub formats: Vec<FormatProgress>,
}

/// Full job progress structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobProgress {
    /// Legacy field for backwards compatibility
    pub current_step: String,
    /// Current processing stage
    pub stage: JobStage,
    /// Total number of sources
    pub total_sources: u64,
    /// Number of processed sources
    pub processed_sources: u64,
    /// Currently processing source (legacy)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_source: Option<String>,
    /// Queue position (for queued jobs)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub queue_position: Option<u64>,
    /// Time remaining in queue delay (ms)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub queue_delay_remaining_ms: Option<u64>,
    /// Per-source progress
    #[serde(default)]
    pub sources: Vec<SourceProgress>,
    /// Whitelist stage progress
    #[serde(skip_serializing_if = "Option::is_none")]
    pub whitelist: Option<WhitelistProgress>,
    /// Generation stage progress
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generation: Option<GenerationProgress>,
    /// When current stage started
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stage_started_at: Option<String>,
}

impl Default for JobProgress {
    fn default() -> Self {
        Self {
            current_step: "queued".to_string(),
            stage: JobStage::Queue,
            total_sources: 0,
            processed_sources: 0,
            current_source: None,
            queue_position: None,
            queue_delay_remaining_ms: None,
            sources: Vec::new(),
            whitelist: None,
            generation: None,
            stage_started_at: None,
        }
    }
}

impl JobProgress {
    /// Create progress for downloading stage
    pub fn downloading(total_sources: u64) -> Self {
        Self {
            current_step: "downloading".to_string(),
            stage: JobStage::Downloading,
            total_sources,
            processed_sources: 0,
            current_source: None,
            queue_position: None,
            queue_delay_remaining_ms: None,
            sources: Vec::new(),
            whitelist: None,
            generation: None,
            stage_started_at: Some(chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S%.6f").to_string()),
        }
    }

    /// Update to whitelist stage
    pub fn to_whitelist(&mut self, domains_before: u64) {
        self.current_step = "whitelist".to_string();
        self.stage = JobStage::Whitelist;
        self.whitelist = Some(WhitelistProgress {
            domains_before,
            domains_after: domains_before,
            total_removed: 0,
            processing: true,
            patterns: Vec::new(),
        });
        self.stage_started_at = Some(chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S%.6f").to_string());
    }

    /// Update to generation stage
    pub fn to_generation(&mut self, total_domains: u64) {
        self.current_step = "generation".to_string();
        self.stage = JobStage::Generation;
        self.generation = Some(GenerationProgress {
            current_format: None,
            formats: vec![
                FormatProgress {
                    format: "hosts".to_string(),
                    status: FormatStatus::Pending,
                    domains_written: 0,
                    total_domains,
                    percent: 0.0,
                    file_size: None,
                    gz_size: None,
                },
                FormatProgress {
                    format: "plain".to_string(),
                    status: FormatStatus::Pending,
                    domains_written: 0,
                    total_domains,
                    percent: 0.0,
                    file_size: None,
                    gz_size: None,
                },
                FormatProgress {
                    format: "adblock".to_string(),
                    status: FormatStatus::Pending,
                    domains_written: 0,
                    total_domains,
                    percent: 0.0,
                    file_size: None,
                    gz_size: None,
                },
            ],
        });
        self.stage_started_at = Some(chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S%.6f").to_string());
    }

    /// Mark as completed
    pub fn to_completed(&mut self) {
        self.current_step = "completed".to_string();
        self.stage = JobStage::Completed;
        self.stage_started_at = Some(chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S%.6f").to_string());
    }
}

/// Job result on completion
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobResult {
    pub sources_processed: u64,
    pub sources_failed: u64,
    pub total_domains: u64,
    pub unique_domains: u64,
    pub whitelisted_removed: u64,
    #[serde(default)]
    pub output_files: Vec<OutputFile>,
    #[serde(default, skip_serializing_if = "std::collections::HashMap::is_empty")]
    pub categories: std::collections::HashMap<String, u64>,
    #[serde(default)]
    pub errors: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skip_reason: Option<String>,
}

/// Output file info
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OutputFile {
    pub name: String,
    pub format: String,
    pub size_bytes: u64,
    pub domain_count: u64,
}

impl JobResult {
    /// Create a success result
    pub fn success(
        sources_processed: u64,
        sources_failed: u64,
        total_domains: u64,
        unique_domains: u64,
        whitelisted_removed: u64,
        output_files: Vec<OutputFile>,
    ) -> Self {
        Self {
            sources_processed,
            sources_failed,
            total_domains,
            unique_domains,
            whitelisted_removed,
            output_files,
            categories: std::collections::HashMap::new(),
            errors: Vec::new(),
            skip_reason: None,
        }
    }

    /// Create a failure result
    pub fn failure(errors: Vec<String>) -> Self {
        Self {
            sources_processed: 0,
            sources_failed: 0,
            total_domains: 0,
            unique_domains: 0,
            whitelisted_removed: 0,
            output_files: Vec::new(),
            categories: std::collections::HashMap::new(),
            errors,
            skip_reason: None,
        }
    }

    /// Create a skip result
    pub fn skipped(reason: String) -> Self {
        Self {
            sources_processed: 0,
            sources_failed: 0,
            total_domains: 0,
            unique_domains: 0,
            whitelisted_removed: 0,
            output_files: Vec::new(),
            categories: std::collections::HashMap::new(),
            errors: Vec::new(),
            skip_reason: Some(reason),
        }
    }
}

File: backend/app/models/progress.py
"""
Progress tracking data classes for job processing.
"""

from dataclasses import dataclass, field, asdict
from datetime import datetime
from typing import Optional, List, Dict, Any


@dataclass
class SourceProgress:
    """Progress tracking for a single source."""

    id: str  # URL hash
    name: str
    url: str
    status: str = "pending"  # pending, downloading, processing, completed, failed
    cache_hit: Optional[bool] = None
    bytes_downloaded: int = 0
    bytes_total: Optional[int] = None
    download_time_ms: Optional[int] = None
    domain_count: Optional[int] = None
    domain_change: Optional[int] = None  # vs previous run
    error: Optional[str] = None
    warnings: List[str] = field(default_factory=list)
    started_at: Optional[str] = None
    completed_at: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary with calculated fields."""
        d = asdict(self)
        # Calculate download percent
        if self.bytes_total and self.bytes_total > 0:
            d["download_percent"] = round(
                self.bytes_downloaded / self.bytes_total * 100, 1
            )
        else:
            d["download_percent"] = None
        return d


@dataclass
class WhitelistPatternProgress:
    """Progress tracking for a whitelist pattern."""

    pattern: str
    pattern_type: str  # exact, wildcard, regex, subdomain
    match_count: int = 0
    samples: List[str] = field(default_factory=list)  # First 5 matched domains

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return asdict(self)


@dataclass
class WhitelistProgress:
    """Progress tracking for whitelist filtering stage."""

    domains_before: int = 0
    domains_after: int = 0
    total_removed: int = 0
    patterns: List[WhitelistPatternProgress] = field(default_factory=list)
    processing: bool = False

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "domains_before": self.domains_before,
            "domains_after": self.domains_after,
            "total_removed": self.total_removed,
            "patterns": [p.to_dict() for p in self.patterns],
            "processing": self.processing,
        }


@dataclass
class FormatProgress:
    """Progress tracking for output format generation."""

    format: str  # hosts, plain, adblock
    status: str = "pending"  # pending, generating, compressing, completed
    domains_written: int = 0
    total_domains: int = 0
    file_size: Optional[int] = None
    gz_size: Optional[int] = None

    @property
    def percent(self) -> int:
        """Calculate completion percentage."""
        if self.total_domains == 0:
            return 0
        return int(self.domains_written / self.total_domains * 100)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary with calculated fields."""
        return {
            "format": self.format,
            "status": self.status,
            "domains_written": self.domains_written,
            "total_domains": self.total_domains,
            "percent": self.percent,
            "file_size": self.file_size,
            "gz_size": self.gz_size,
        }


@dataclass
class GenerationProgress:
    """Progress tracking for output generation stage."""

    formats: List[FormatProgress] = field(default_factory=list)
    current_format: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "formats": [f.to_dict() for f in self.formats],
            "current_format": self.current_format,
        }


@dataclass
class EnhancedProgress:
    """Complete enhanced job progress tracking."""

    stage: str = "queue"  # queue, downloading, whitelist, generation, completed

    # Queue stage
    queue_position: Optional[int] = None
    queue_delay_remaining_ms: Optional[int] = None

    # Downloading stage
    total_sources: int = 0
    processed_sources: int = 0
    sources: List[SourceProgress] = field(default_factory=list)

    # Whitelist stage
    whitelist: Optional[WhitelistProgress] = None

    # Generation stage
    generation: Optional[GenerationProgress] = None

    # Timing
    stage_started_at: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for Socket.IO emission."""
        return {
            "stage": self.stage,
            "queue_position": self.queue_position,
            "queue_delay_remaining_ms": self.queue_delay_remaining_ms,
            "total_sources": self.total_sources,
            "processed_sources": self.processed_sources,
            "sources": [s.to_dict() for s in self.sources],
            "whitelist": self.whitelist.to_dict() if self.whitelist else None,
            "generation": self.generation.to_dict() if self.generation else None,
            "stage_started_at": self.stage_started_at,
            # Legacy compatibility fields
            "current_step": self.stage,
            "current_source": self._get_current_source(),
        }

    def _get_current_source(self) -> Optional[str]:
        """Get the name of the currently processing source for legacy compatibility."""
        for source in self.sources:
            if source.status in ("downloading", "processing"):
                return source.name
        return None

    def get_source_by_id(self, source_id: str) -> Optional[SourceProgress]:
        """Get source progress by ID."""
        for source in self.sources:
            if source.id == source_id:
                return source
        return None

    def update_source(self, source: SourceProgress) -> None:
        """Update a source's progress."""
        for i, s in enumerate(self.sources):
            if s.id == source.id:
                self.sources[i] = source
                return
        # Source not found, add it
        self.sources.append(source)

    def count_completed_sources(self) -> int:
        """Count sources with completed or failed status."""
        return sum(1 for s in self.sources if s.status in ("completed", "failed"))

    def set_stage(self, stage: str) -> None:
        """Update the current stage with timestamp."""
        self.stage = stage
        self.stage_started_at = datetime.utcnow().isoformat()
